pkg/storage/legacysql/dualwrite/dualwriter.go

```diff
@@ -0,0 +1,229 @@
+package dualwrite
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	"github.com/grafana/grafana-app-sdk/logging"
+	grafanarest "github.com/grafana/grafana/pkg/apiserver/rest"
+)
+
+var (
+	_ grafanarest.Storage = (*dualWriter)(nil)
+)
+
+// dualWriter will write first to legacy, then to unified keeping the same internal ID
+type dualWriter struct {
```

Comment:
This struct combines the logic from `mode1`,`mode2`, and `mode3` into one path controlled by two flags: readUnified (mode3) and errorIsOK (mode1)